rpc/RPCTransaction.java

```diff
@@ -135,45 +129,159 @@ public void close() {
+                if (collector.isDone()) collectors.remove(requestId);
+            }
+
+            @Override
+            public void onError(Throwable error) {
+                assert error instanceof StatusRuntimeException : "The server only yields these exceptions";
+                final Response res = new Response.Error((StatusRuntimeException) error);
+                // TODO: is it desirable behaviour to kill all the pending requests?
+                // TODO: this isn't nice in any case - the error for other pending requests is misleading
+                collectors.values().parallelStream().forEach(x -> x.onResponse(res));
+                collectors.clear();
+                close();
+            }
+
+            @Override
+            public void onCompleted() {
+                final Response res = new Response.Error(CONNECTION_CLOSED);
+                collectors.values().parallelStream().forEach(x -> x.onResponse(res));
+                collectors.clear();
+                // TODO: maybe this should just be close()? But if the server terminates abruptly...
+                isOpen.set(false);
+            }
+        };
     }
 
-    private void processResponse(final RPCResponse res) {
-        // TODO: This probably doesn't work if we send a slow query followed by a fast query down the same transaction.
-        final RPCResponseAccumulator currentCollector = collectorQueue.peek();
-        assert currentCollector != null;
-        currentCollector.onResponse(res);
-        if (currentCollector.isDone()) collectorQueue.poll();
+    abstract static class Response {
+
+        abstract TransactionProto.Transaction.Res read();
+
+        static class Ok extends Response {
+            private final TransactionProto.Transaction.Res response;
+
+            Ok(TransactionProto.Transaction.Res response) {
+                this.response = response;
+            }
+
+            @Override
+            TransactionProto.Transaction.Res read() {
+                return response;
+            }
+
+            @Override
+            public String toString() {
+                return className(getClass()) + "{" + response + "}";
+            }
+        }
+
+        static class Error extends Response {
+            private final GraknClientException error;
+
+            Error(StatusRuntimeException error) {
+                // TODO: parse different gRPC errors into specific GraknClientException
+                this.error = new GraknClientException(error);
+            }
+
+            Error(ErrorMessage error) {
+                this.error = new GraknClientException(error);
+            }
+
+            @Override
+            TransactionProto.Transaction.Res read() {
+                throw error;
+            }
+
+            @Override
+            public String toString() {
+                return className(getClass()) + "{" + error + "}";
+            }
+        }
+    }
+
+    abstract static class ResponseCollector {
```

Comment:
I'm struggling with naming this class. Is it a `ResponseCollector`? A `ResponseBuffer`? A `ResponseBucket`? A `ResponseQueue`?

Conceptually, we maintain a map of Requests to Response buckets. Whenever a Response comes in it gets added to the correct bucket.

Internally, the bucket is implemented as a response _buffer_. Theoretically, we could limit the buffer size to perform flow control.

But then comes the issue of differentiating a Single response collector (that will only ever collect one response) and a Multiple response collector (that may receive multiple responses). What is a `ResponseBuffer.Single` supposed to mean? In this case `ResponseCollector.Single` seems the clearest. How many responses do we collect? Just a single one.

Reply:
I'm starting to feel that the `ResponseCollector` is correct - "collecting" has a notion of collecting something into something else already. I think the API would be a bit clearer if you rename the `put` method to `collect`. Then you get:
```
collector.collect(response
```
vs
```
collector.put(response)
```
i think the former has more story to it.

alternatives we might consider are also `collector.record(response)` or `collector.buffer(response)`, which convey more of the semantic of saving responses somewhere to be used later!

Reply:
I think if its called `ResponseCollector` it should actually be `.collect()`. if its `ResponseBuffer` it shuold be `.buffer()` etc. 

overall collector seems about right

Reply:
Hmmm. I don't really like `collect`. `ResponseCollector` roughly describes the big picture of what the thing does over time, over the course of multiple `put`s. I want a name that makes it clear that this is adding a single response to the collector.

The old name was `onResponse`, but `add` seems good, `put` too, `receive` would be OK.

Reply:
We settled on `add` in the end.